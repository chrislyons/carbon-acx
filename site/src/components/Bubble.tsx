import { useMemo } from 'react';

import { formatEmission, formatKilograms } from '../lib/format';
import {
  formatReferenceHint,
  ReferenceCarrier,
  ReferenceLookup,
  resolveReferenceIndices
} from '../lib/references';

export interface BubbleDatum extends ReferenceCarrier {
  layer_id?: string | null;
  activity_id?: string | null;
  activity_name?: string | null;
  category?: string | null;
  sector?: string | null;
  values?: {
    mean?: number | null;
  } | null;
  units?: Record<string, string | null> | null;
}

export interface BubbleProps {
  title?: string;
  data?: BubbleDatum[] | null;
  referenceLookup: ReferenceLookup;
  variant?: 'card' | 'embedded';
}

interface BubblePoint {
  key: string;
  label: string;
  category: string;
  grams: number;
  kilograms: number;
  hint: string;
}

const SVG_WIDTH = 640;
const SVG_HEIGHT = 360;
const PADDING_X = 80;
const PADDING_Y = 50;
const MAX_RADIUS = 42;
const AXIS_FONT_SIZE = 11;
const AXIS_TICK_FONT_SIZE = 10;
function toCategory(value: string | null | undefined): string {
  if (!value) {
    return 'Other';
  }
  return value;
}

export function Bubble({
  title = 'Activity emissions bubble chart',
  data,
  referenceLookup,
  variant = 'card'
}: BubbleProps) {
  const points = useMemo<BubblePoint[]>(() => {
    if (!Array.isArray(data)) {
      return [];
    }
    return data
      .map((row, index) => {
        const mean = typeof row?.values?.mean === 'number' ? row.values.mean : null;
        if (mean == null || !Number.isFinite(mean) || mean <= 0) {
          return null;
        }
        const kilograms = mean / 1_000;
        const label = row?.activity_name || row?.activity_id || `Activity ${index + 1}`;
        const indices = resolveReferenceIndices(row, referenceLookup);
        return {
          key: `${row?.activity_id ?? index}-${index}`,
          label,
          category: toCategory(row?.category),
          grams: mean,
          kilograms,
          hint: `${label} — ${formatEmission(mean)} ${formatReferenceHint(indices)}`
        } satisfies BubblePoint;
      })
      .filter((value): value is BubblePoint => value !== null)
      .sort((a, b) => b.grams - a.grams);
  }, [data, referenceLookup]);

  const categories = useMemo(
    () => Array.from(new Set(points.map((point) => point.category))),
    [points]
  );

  const maxKg = useMemo(() => points.reduce((max, point) => Math.max(max, point.kilograms), 0), [points]);

  if (points.length === 0 || maxKg <= 0) {
    if (variant === 'card') {
      return (
        <section
          aria-labelledby="bubble-heading"
          className="rounded-2xl border border-slate-800/80 bg-slate-950/60 p-5 shadow-inner shadow-slate-900/40"
          id="bubble"
          role="region"
          tabIndex={-1}
        >
          <h3 id="bubble-heading" className="text-base font-semibold text-slate-100">
            {title}
          </h3>
          <p className="mt-4 text-sm text-slate-400">No activity data available.</p>
        </section>
      );
    }
    return <p className="text-sm text-slate-400">No activity data available.</p>;
  }

  const chartWidth = SVG_WIDTH - PADDING_X * 2;
  const chartHeight = SVG_HEIGHT - PADDING_Y * 2;
  const xStep = chartWidth / Math.max(categories.length, 1);

  const ticks = [0, maxKg / 2, maxKg];

  const chart = (
    <div className="mt-4 flex flex-col items-stretch">
      <svg
        data-testid="bubble-svg"
        role="img"
        viewBox={`0 0 ${SVG_WIDTH} ${SVG_HEIGHT}`}
        className="w-full text-slate-400"
        aria-describedby="bubble-axis-description"
      >
          <defs>
            <radialGradient id="bubble-fill" cx="50%" cy="50%" r="75%">
              <stop offset="0%" stopColor="#38bdf8" stopOpacity="0.9" />
              <stop offset="100%" stopColor="#0f172a" stopOpacity="0.2" />
            </radialGradient>
          </defs>
          <rect
            x={PADDING_X}
            y={PADDING_Y}
            width={chartWidth}
            height={chartHeight}
            fill="url(#bubble-fill)"
            fillOpacity={0.08}
            stroke="rgba(148, 163, 184, 0.25)"
            rx={24}
          />
          {points.map((point, index) => {
            const categoryIndex = categories.indexOf(point.category);
            const cx = PADDING_X + xStep * (categoryIndex + 0.5);
            const relative = point.kilograms / maxKg;
            const cy = PADDING_Y + chartHeight - relative * chartHeight;
            const radius = Math.max(8, Math.sqrt(relative) * MAX_RADIUS);
            return (
              <g
                key={point.key}
                className="group"
                transform={`translate(${cx}, ${cy})`}
                data-testid={`bubble-point-${index}`}
              >
                <circle
                  r={radius}
                  className="fill-sky-400/80 transition-transform duration-300 ease-out group-hover:scale-110 group-focus-within:scale-110"
                  stroke="rgba(56, 189, 248, 0.4)"
                  strokeWidth={2}
                >
                  <title>{point.hint}</title>
                </circle>
              </g>
            );
          })}
          {categories.map((category, index) => {
            const cx = PADDING_X + xStep * (index + 0.5);
            return (
              <text
                key={category}
                x={cx}
                y={SVG_HEIGHT - 12}
                textAnchor="middle"
                className="fill-slate-400"
                style={{ fontSize: `${AXIS_TICK_FONT_SIZE}px` }}
              >
                {category}
              </text>
            );
          })}
          {ticks.map((tick) => {
            const relative = tick / maxKg;
            const y = PADDING_Y + chartHeight - relative * chartHeight;
            return (
              <g key={`tick-${tick}`}>
                <line
                  x1={PADDING_X - 8}
                  x2={PADDING_X}
                  y1={y}
                  y2={y}
                  stroke="rgba(148, 163, 184, 0.4)"
                />
                <text
                  x={PADDING_X - 12}
                  y={y + 4}
                  textAnchor="end"
                  className="fill-slate-400"
                  style={{ fontSize: `${AXIS_TICK_FONT_SIZE}px` }}
                >
                  {formatKilograms(tick)}
                </text>
              </g>
            );
          })}
          <text
            id="bubble-axis-description"
            x={PADDING_X - 48}
            y={SVG_HEIGHT / 2}
            textAnchor="middle"
            className="fill-slate-500"
            style={{ fontSize: `${AXIS_FONT_SIZE}px` }}
            transform={`rotate(-90 ${PADDING_X - 48} ${SVG_HEIGHT / 2})`}
          >
            Annual emissions (kg CO₂e)
          </text>
          <text
            x={SVG_WIDTH / 2}
            y={SVG_HEIGHT - 4}
            textAnchor="middle"
            className="fill-slate-500"
            style={{ fontSize: `${AXIS_FONT_SIZE}px` }}
          >
            Activity category
          </text>
      </svg>
    </div>
  );

  if (variant === 'card') {
    return (
      <section
        aria-labelledby="bubble-heading"
        className="rounded-2xl border border-slate-800/80 bg-slate-950/60 p-5 shadow-inner shadow-slate-900/40"
        id="bubble"
        role="region"
        tabIndex={-1}
      >
        <h3 id="bubble-heading" className="text-base font-semibold text-slate-100">
          {title}
        </h3>
        {chart}
      </section>
    );
  }

  return chart;
}
