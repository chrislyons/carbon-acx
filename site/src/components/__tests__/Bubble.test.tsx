import '@testing-library/jest-dom';
import { render } from '@testing-library/react';

import { Bubble } from '../Bubble';

const referenceLookup = new Map<string, number>([
  ['SRC.A', 1]
]);

const sampleData = [
  {
    activity_id: 'cooling',
    activity_name: 'Cooling',
    category: 'HVAC',
    values: { mean: 2500 },
    citation_keys: ['SRC.A']
  },
  {
    activity_id: 'lighting',
    activity_name: 'Lighting',
    category: 'Facilities',
    values: { mean: 1200 },
    hover_reference_indices: [2]
  }
];

describe('Bubble', () => {
  it('renders bubble chart with pulsing circles', () => {
    const { getByRole, getByTestId } = render(
      <Bubble data={sampleData} referenceLookup={referenceLookup} />
    );

    // Section is a landmark with an accessible name and focus target.
    const section = getByRole('region', { name: /bubble/i });
    expect(section).toHaveAttribute('tabindex', '-1');

    const svg = getByTestId('bubble-svg');
    expect(svg).toMatchSnapshot();
    const bubble = getByTestId('bubble-point-0');
    const title = bubble.querySelector('title');
    expect(title?.textContent).toContain('[1]');
  });
});
