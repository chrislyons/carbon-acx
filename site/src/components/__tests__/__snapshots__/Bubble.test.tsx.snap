// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Bubble > renders bubble chart with pulsing circles 1`] = `
<svg
  aria-describedby="bubble-axis-description"
  class="w-full text-slate-400"
  data-testid="bubble-svg"
  role="img"
  viewBox="0 0 640 480"
>
  <defs>
    <radialgradient
      cx="50%"
      cy="50%"
      id="bubble-fill"
      r="75%"
    >
      <stop
        offset="0%"
        stop-color="#38bdf8"
        stop-opacity="0.9"
      />
      <stop
        offset="100%"
        stop-color="#0f172a"
        stop-opacity="0.2"
      />
    </radialgradient>
  </defs>
  <rect
    fill="url(#bubble-fill)"
    fill-opacity="0.08"
    height="380"
    rx="24"
    stroke="rgba(148, 163, 184, 0.25)"
    width="480"
    x="80"
    y="50"
  />
  <g
    class="group"
    data-testid="bubble-point-0"
    transform="translate(200, 50)"
  >
    <circle
      class="fill-sky-400/80 transition-transform duration-300 ease-out group-hover:scale-110 group-focus-within:scale-110"
      r="42"
      stroke="rgba(56, 189, 248, 0.4)"
      stroke-width="2"
    >
      <title>
        Cooling — 2.50 kg CO₂e [1]
      </title>
    </circle>
  </g>
  <g
    class="group"
    data-testid="bubble-point-1"
    transform="translate(440, 247.6)"
  >
    <circle
      class="fill-sky-400/80 transition-transform duration-300 ease-out group-hover:scale-110 group-focus-within:scale-110"
      r="29.09845356715714"
      stroke="rgba(56, 189, 248, 0.4)"
      stroke-width="2"
    >
      <title>
        Lighting — 1.20 kg CO₂e [2]
      </title>
    </circle>
  </g>
  <text
    class="fill-slate-400"
    style="font-size: 10px;"
    text-anchor="middle"
    x="200"
    y="468"
  >
    HVAC
  </text>
  <text
    class="fill-slate-400"
    style="font-size: 10px;"
    text-anchor="middle"
    x="440"
    y="468"
  >
    Facilities
  </text>
  <g>
    <line
      stroke="rgba(148, 163, 184, 0.4)"
      x1="72"
      x2="80"
      y1="430"
      y2="430"
    />
    <text
      class="fill-slate-400"
      style="font-size: 10px;"
      text-anchor="end"
      x="68"
      y="434"
    >
      0.0 kg
    </text>
  </g>
  <g>
    <line
      stroke="rgba(148, 163, 184, 0.4)"
      x1="72"
      x2="80"
      y1="240"
      y2="240"
    />
    <text
      class="fill-slate-400"
      style="font-size: 10px;"
      text-anchor="end"
      x="68"
      y="244"
    >
      1.3 kg
    </text>
  </g>
  <g>
    <line
      stroke="rgba(148, 163, 184, 0.4)"
      x1="72"
      x2="80"
      y1="50"
      y2="50"
    />
    <text
      class="fill-slate-400"
      style="font-size: 10px;"
      text-anchor="end"
      x="68"
      y="54"
    >
      2.5 kg
    </text>
  </g>
  <text
    class="fill-slate-500"
    id="bubble-axis-description"
    style="font-size: 11px;"
    text-anchor="middle"
    transform="rotate(-90 32 240)"
    x="32"
    y="240"
  >
    Annual emissions (kg CO₂e)
  </text>
  <text
    class="fill-slate-500"
    style="font-size: 11px;"
    text-anchor="middle"
    x="320"
    y="476"
  >
    Activity category
  </text>
</svg>
`;
